crosswalks/ogcapi-coverages.md

```diff
@@ -0,0 +1,428 @@
+
+This crosswalk will proceed by comparing such standard with the openEO API, touching
+all the functionalities that are declared in the OGC standard, i.e.:
+
+- (meta)data retrieval
+- subsetting through dimension(s)
+- range (bands) subsetting
+- up/down-scaling
+- data retrieval as tiles
+
+NOTE: the following abbreviations will be used throughout the document:
+
+- **openEO** : *openEO API, v1.2.0*
+- **OAC1** : *OGC API - Coverages - Part 1*
+
+## Nomenclature
+
+Despite stemming from the [OGC API - Features](http://docs.opengeospatial.org/is/17-069r3/17-069r3.html) standards,
+openEO/STAC and OGC have both a different terminology and data models, so here
+we put the definitions of the most relevant terms of both "worlds".
+
+Most notably, an OGC [*coverage*](http://myogc.org/go/coveragesDWG) &mdash;
+even though formally defined as a *feature*
+(and not a *collection* thereof) &mdash; is frequently an abstraction for a 
+*collection* of (generally, but not limited to, gridded) datasets/items
+(like time-series of satellite images for instance),
+so we can generally compare it with STAC [collections](https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md),
+especially [datacubes](https://openeo.org/documentation/1.0/datacubes.html).
+
+### OGC
+
+**`collection`**
+A set of **features** from a dataset.
+
+**`coverage`**
+A **feature** that acts as a function to return values from its **range** for any direct
+position within its spatial, temporal or spatio-temporal **domain**;
+this can be equivalent to either a [STAC Item](https://github.com/radiantearth/stac-spec/tree/master/item-spec)
+or [Collection](https://github.com/radiantearth/stac-spec/blob/master/collection-spec/collection-spec.md).
+
+**`domain`** 
+A set of direct positions along some *axis* in space and time (or other dimensions) that define the
+"location" of a coverage, associated with a (compounding of) spatio(-temporal)(-other) coordinates
+reference system (CRS), and for each of which a set of 1+ values (**range**) are associated.
+
+**`feature`**
+An abstraction of real world phenomena.
+
+**`range`**
+Set of **feature** attribute values associated by a function with the elements of the **domain** of a 
+coverage, that means the stored values of a coverage like for instance the spectral bands;
+in the STAC [`datacube`](https://github.com/stac-extensions/datacube) extension, **range** and **domain**
+are condensed into the same **cube:dimensions** container.
+
+### STAC
+
+**`catalog`**
+A logical group of other **catalog**, **collection**, and **item** objects.
+
+**`collection`**
+A set of common fields to describe a group of **items** that share properties and metadata
+
+**`item`**
+A GeoJSON Feature augmented with foreign members relevant to a STAC object.
+
+**`cube:dimensions`**
+A set of dimensions that can be of different types: spatial, temporal, "other", etc.
+
+
+## Crosswalk
+
+*Executive summary (tl;dr)*
```

Comment:
For me an important details is missing here. Can OGC API - Coverages be combined with the openEO API? It seems they have a common baseline and it could just be additional endpoints (mostly `/collections/{collectionId}/coverage/...`). 

Due to the openEO WCPS driver, we could translate openEO to WCPS, right? It seems vice-versa is not easily possible (waiting for an answer from Peter Baumann).

Reply:
There are probably a couple of assumptions to be made regarding the cubes' dimensions, which makes openEO/OGC not 100% compatible, but probably cover most of the "real" coverages that are there anyway: I mean that there is not necessarily a 1:1 agreement and directional alignment between the dimensions of a cube and the CRS' axes (imagine I don't know, an oblique plane/tile in a 3D fishtank), which is a case that the OGC Coverage model can describe, while I am not sure about this with openEO. ?

Also, since the `datacube` STAC extension merges a coverage's *domain* and *range* spaces into the same "dimensions" container, there is the little "pain" to define the way to identify which dimensions of a datacube are what (spatial, temporal, other, band, etc), based on fragile assumptions on axis/dimensions labels?

Translating a (subset of) (the most basic, like filtering, scaling, etc) openEO processes to WCPS is feasible (are you talking about the [openEO Spring driver](https://github.com/Open-EO/openeo-spring-driver)?).

The other way around is probably also feasible, with the additional step of requiring to identify the category of the dimensions, as openEO processes are much more strongly typed in this sense, WCPS treats all dimensions equally and operations on them do not "discriminate".

Reply:
I guess openEO could describe the 3D fishtank if there's a CRS for it?

I agree that the STAC data cube extension is a bit messy with regards to the type, but it's specified. type = spatial (with axis), type = temporal, type = ... (e.g. bands). At least internally it should be clear. "other" only really shows up if you add a "user-defined" dimension in the process graph.

I'm not specifically talking about any implementation, more what theoretically can be translated or whether there are any roadblocks.

openEO processes only really "discriminate" spatial axes as x and y need usually be used together, but I'm not sure how WCPS handles the special case of spatial?

Reply:
> I guess openEO could describe the 3D fishtank if there's a CRS for it?

Yes, but how can the orientation of the cube dimensions be specified? Even in the simple/common case when the cube is "aligned" with the CRS axes: we should not rely on the labels of the dimensions to infer its "direction". OGC uses the "offset vectors" for this.

> I agree that the STAC data cube extension is a bit messy with regards to the type, but it's specified. type = spatial (with axis), type = temporal, type = ... (e.g. bands). At least internally it should be clear. "other" only really shows up if you add a "user-defined" dimension in the process graph.

I don't think it is messy, it is clear! Only I personally prefer the distinction that the OGC Coverages do between domain and range, it makes it even clearer.

> openEO processes only really "discriminate" spatial axes as x and y need usually be used together, but I'm not sure how WCPS handles the special case of spatial?

Afaik WCPS does not handle any special spatial (nor temporal) case: all operators are completely agnostic of the type of dimensions they operate onto.

Reply:
Dimension labels in openEO have an order, but otherwise there's nothing like that, indeed.

> Afaik WCPS does not handle any special spatial (nor temporal) case

Well, it does, right? bands (range) seems separate from spatial/temporal (domain), while in openEO it's just all a dimension. What's so special about bands?

Reply:
Yep, I was talking indeed about a Coverage dimensions: no special case there, but yes, bands indeed are treated separately. To me it makes a whole lot of sense to have a domain, a "space", (or space-time) **where** the measurements lie, and then you have the data-space, which is **what** are the measurements about.

Reply:
Okay, it's good to understand this, thanks.